app/src/main/java/org/mozilla/tv/firefox/webrender/cursor/CursorModel.kt

```diff
@@ -0,0 +1,352 @@
+        }
+        when (event.action) {
+            KeyEvent.ACTION_UP -> directionKeysPressed -= direction
+            KeyEvent.ACTION_DOWN -> {
+                directionKeysPressed += direction
+                pushCursorEvent(direction)
+            }
+            else -> return false
+        }
+
+        _isCursorMoving.onNext(!directionKeysPressed.isEmpty())
+
+        return true
+    }
+
+    private fun pushCursorEvent(direction: Direction) {
+        val edgeNearCursor = getEdgeOfScreenNearCursor()
+        val couldScroll = edgeNearCursor?.let { webViewCouldScrollInDirectionProvider(it) }
+
+        val cursorMovedToEdgeOfScreen = edgeNearCursor == direction
+        val endOfDomContentReached = couldScroll == false
+
+        val event = if (cursorMovedToEdgeOfScreen && endOfDomContentReached) {
+            CursorEvent.ScrolledToEdge(direction)
+        } else {
+            CursorEvent.CursorMoved(direction)
+        }
+
+        _cursorMovedEvents.onNext(event)
+    }
+
+    /**
+     * @return returns a MotionEvent if the event is consumed, null otherwise
+     */
+    @SuppressLint("Recycle")
+    @CheckResult(suggest = "Recycle MotionEvent after use")
+    private fun maybeToMotionEvent(event: KeyEvent): MotionEvent? {
+        if (!isCursorActive.blockingFirst()) {
+            return null
+        }
+        if (event.keyCode != KeyEvent.KEYCODE_DPAD_CENTER && event.keyCode != KeyEvent.KEYCODE_ENTER) {
+            return null
+        }
+
+        val motionEvent = MotionEvent.obtain(event.downTime, event.eventTime, event.action,
+                lastKnownCursorPos.x, lastKnownCursorPos.y, 0)
+        return when (event.action) {
+            KeyEvent.ACTION_UP -> {
+                _isSelectPressed.onNext(false)
+                motionEvent
+            }
+            KeyEvent.ACTION_DOWN -> {
+                _isSelectPressed.onNext(true)
+                motionEvent
+            }
+            else -> null
+        }
+    }
+
+    /**
+     * Called by CursorView in order to drive cursor movement.
+     *
+     * Performance:
+     * Previous code updated cursor VM position every 16MS (1_000 MS / 60 frames == 16.66). This
+     * led to minor jitter because we were not tied exactly to the actual framerate. This new
+     * implementation is harder to understand, but is noticeably more performant.
+     *
+     * Architecture:
+     * Having a View#onDraw call VM updates does not mesh well with our architecture, but we found
+     * no way to achieve desired performance with any other approach. Instead, we contained this
+     * architectural aberration to this class, and as much as possible it should be left here.
+     *
+     * STRONGLY avoid accessing this state outside of this class, and think very carefully about
+     * how you do if it is unavoidable.
+     *
+     * @return whether or not the view should continue to invalidate itself
+     */
+    fun mutatePosition(oldPos: PointF): Boolean {
```

Comment:
nit: `oldPos` is not quite correct. Maybe `oldPosAndReturnedPos`.

It might even be clearer to have two: `fun mutatePosition(oldPos: PointF, returnedPosition: PointF)`